plonk-hashing/src/poseidon/round_constant.rs

```diff
@@ -0,0 +1,171 @@
+use alloc::collections::vec_deque::VecDeque;
+use ark_ff::{BigInteger, PrimeField};
+/// From the paper
+/// THe parameter describes the initial state of constant generation (80-bits)
+/// * `field`: description of field. b0, b1
+/// * `sbox`: description of s-box. b2..=b5
+/// * `field_size`: binary representation of field size. b6..=b17
+/// * `t`: binary representation of t. b18..=b29
+/// * `rf`: binary representation of rf. b30..=b39
+/// * `rp`: binary representation of rp. b40..=b49
+/// * `ones`: set to 1. b50..=b79
+pub fn generate_constants<F: PrimeField>(
+    field: u8,
+    sbox: u8,
+    field_size: u16,
+    t: u16,
+    r_f: u16,
+    r_p: u16,
+) -> Vec<F> {
+    let n_bytes = (F::size_in_bits() + 8 - 1) / 8;
+    if n_bytes != 32 {
+        unimplemented!("neptune currently supports 32-byte fields exclusively");
+    };
+    assert_eq!((field_size as f32 / 8.0).ceil() as usize, n_bytes);
+
+    // r_f here is 2* number of *half* full rounds.
+    let num_constants = (r_f + r_p) * t;
+    let mut init_sequence: VecDeque<bool> = VecDeque::new();
+    append_bits(&mut init_sequence, 2, field); // Bits 0-1
+    append_bits(&mut init_sequence, 4, sbox); // Bits 2-5
+    append_bits(&mut init_sequence, 12, field_size); // Bits 6-17
+    append_bits(&mut init_sequence, 12, t); // Bits 18-29
+    append_bits(&mut init_sequence, 10, r_f); // Bits 30-39
+    append_bits(&mut init_sequence, 10, r_p); // Bits 40-49
+    append_bits(&mut init_sequence, 30, 0b111111111111111111111111111111u128); // Bits 50-79
+
+    let mut grain = GrainLFSR::new(init_sequence, field_size);
+    let mut round_constants: Vec<F> = Vec::new();
+
+    match field {
+        1 => {
+            for _ in 0..num_constants {
+                while {
```

Comment:
use `F::from_le_bytes_mod_order` instead of `while` loop here. See https://github.com/arkworks-rs/sponge/blob/51d6fc9aac1fa69f44a04839202b5de828584ed8/src/poseidon/grain_lfsr.rs#L135